pylint/checkers/refactoring.py

```diff
@@ -486,24 +486,28 @@ def visit_call(self, node):
     @utils.check_messages('consider-using-enumerate')
     def visit_for(self, node):
         """Emit a convention whenever range and len are used for indexing."""
-        # Verify that we have a `range(len(...))` call and that the object
-        # which is iterated is used as a subscript in the body of the for.
+        # Verify that we have a `range([start], len(...), [stop])` call and
+        # that the object which is iterated is used as a subscript in the
+        # body of the for.
 
         # Is it a proper range call?
         if not isinstance(node.iter, astroid.Call):
             return
         if not self._is_builtin(node.iter.func, 'range'):
             return
-        if len(node.iter.args) != 1:
+        if len(node.iter.args) == 2 and not _is_constant_zero(node.iter.args[0]):
+            return
+        if len(node.iter.args) > 2:
             return
 
         # Is it a proper len call?
-        if not isinstance(node.iter.args[0], astroid.Call):
+        possible_len_possition = len(node.iter.args) - 1
```

Comment:
This on the first sight seems like it's a little too verbose, using `-1` in subscript in places where `0` have been used before a patch is more concise and IMHO is not much harder to comprehend for code reader.

If we really want to introduce new local variable, a better idea may be to do something along of lines of `upper_bound_arg = node.iter.args[-1]` - we have one level less of indirection (using objects themselves instead of indices) and we can see on first sight that we want to analyze upper bound and not some ambiguous len call.